syntax = "proto3";

package plushie_tycoon;
//todo: map in proto is not ordered. Perhaps switch to 2 repeated fields?
//or alternatively we can provide an order object somewhere
service UITransfer {
    rpc buy(TransactionObject) returns (UserOutput) {};
    rpc sell(TransactionObject) returns (UserOutput) {};
    rpc make(TransactionObject) returns (UserOutput) {};
    rpc next(SelectionObject) returns (UserOutput) {};
    rpc save(SelectionObject) returns (UserOutput) {};
    rpc load(SelectionObject) returns (UserOutput) {};
    rpc back(SelectionObject) returns (UserOutput) {};
    rpc quit(SelectionObject) returns (UserOutput) {};
    rpc init(SelectionObject) returns (UserOutput) {};
}


message SelectionObject {
}

message TransactionObject {
    string name = 1;
    int32 quantity = 2;
}
//this should only be reportable metrics to frontend, not everything

message mItemCost {
    double movein = 3;
    double moveout = 4;
    double storage = 5;
}

message mRatioPerProduct {
    map<string, int32> ratio = 6;
}

message UserOutput {
    map<string, int32> prices = 1;
    map<string, int32> quantities = 2;
    map<string, double> weights = 13;
    map<string, double> volumes = 14;
    map<string, mItemCost> item_cost = 12;
    map<string, mRatioPerProduct> resource_ratio = 7;
    string console_output = 8;
    int32 budget = 9;
    int32 time = 10;
    string action = 11;
}